netket/optimizer/sr/sr_onthefly.py

```diff
@@ -0,0 +1,134 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Callable, Optional, Union, Tuple, Any
+from functools import partial
+
+import jax
+import flax
+from jax import numpy as jnp
+from flax import struct
+
+from netket.utils import rename_class
+
+from .sr_onthefly_logic import mat_vec as mat_vec_onthefly, tree_cast
+
+from .base import SR
+
+Ndarray = Any
+
+
+@rename_class(r"SR{Onthefly}")
+@struct.dataclass
+class SR_otf(SR):
+    """
+    Base class holding the parameters for the iterative SR on the fly method.
+    """
+
+    tol: float = 1.0e-5
+    atol: float = 0.0
+    maxiter: int = None
+    M: Optional[Union[Callable, Ndarray]] = None
+
+    def create(self, *args, **kwargs):
+        return LazySMatrix(*args, **kwargs)
+
+
+@rename_class(r"SR{Onthefly,CG}")
+@struct.dataclass
+class SR_otf_cg(SR_otf):
+    ...
+
+    def solve_fun(self):
+        return partial(
+            jax.scipy.sparse.linalg.cg,
+            tol=self.tol,
+            atol=self.atol,
+            maxiter=self.maxiter,
+            M=self.M,
+        )
+
+
+@rename_class(r"SR{Onthefly,GMRES}")
+@struct.dataclass
+class SR_otf_gmres(SR_otf):
+    restart: int = 20
+    solve_method: str = struct.field(pytree_node=False, default="batched")
+
+    def solve_fun(self):
+        return partial(
+            jax.scipy.sparse.linalg.gmres,
+            tol=self.tol,
+            atol=self.atol,
+            maxiter=self.maxiter,
+            M=self.M,
+            restart=self.restart,
+            solve_method=self.solve_method,
+        )
+
+
+@partial(jax.jit, static_argnums=0)
+def apply_onthefly(apply_fun, params, samples, model_state, sr, grad, x0):
+    # Preapply the model state so that when computing gradient we only
+    # get gradient of parameeters
+    def fun(W, σ):
+        return apply_fun({"params": W, **model_state}, σ)
+
+    grad = tree_cast(grad, params)
+    # we could cachee this...
+    if x0 is None:
+        x0 = jax.tree_map(jnp.zeros_like, grad)
+
+    _mat_vec = partial(
+        mat_vec_onthefly,
+        forward_fn=fun,
+        params=params,
+        samples=samples,
+        diag_shift=sr.diag_shift,
+    )
+    solve_fun = sr.solve_fun()
+    out, _ = solve_fun(_mat_vec, grad, x0=x0)
+    return out
+
+
+@struct.dataclass
+class LazySMatrix:
+    apply_fun: Callable
+    params: Any
+    samples: Any
+    sr: SR_otf
+    model_state: Any = None
+    x0: Any = None
+
+    def __call__(self, grad):
+        return self.apply(grad)
+
+    def apply(self, grad, x0=None):
+        if jnp.ndim(self.samples) != 2:
+            samples = self.samples.reshape((-1, self.samples.shape[-1]))
+
+        if x0 is None:
+            x0 = self.x0
+
+        out = apply_onthefly(
+            self.apply_fun,
+            self.params,
+            samples,
+            self.model_state,
+            self.sr,
+            grad,
+            x0,
+        )
+
+        return out
```

Comment:
I noticed that, given `S = mcstate.quantum_geometric_tensor(sr)`, the EOM `Sẋ = F` is solved by calling
```python
ẋ = S(F)
```
for example here:
https://github.com/netket/netket/blob/2abcc3862e5717d8171c9ab8de5fd84fa1b52008/netket/drivers/vmc.py#L81-L83
I think this is confusing, since this operation is solving a linear equation and mathematically corresponds to applying the _pseudoinverse_ of `S` to `F`, not `S` itself; so it's more like `ẋ = S^-1 F`.

I would suggest to either
1. rename `apply`/`__call__` to `solve`, so we have `ẋ = S.solve(F)` or, alternatively,
2. rename `LazySMatrix` and `.quantym_geoemtric_tensor` so that it becomes clear that this returns an object for solving the SR equation, not the actual quantum Fisher matrix / QGT.

I would prefer 1. and also note that we should add a way to easily get the full QFM if one so desires, as it can be useful to have sometimes for analysis purposes. (The second part can be done independently of this PR, though.)

Reply:
Yes good catch @femtobit this should not be a __call__ method but rather a solve one.

Reply:
Ah yeah, it was on my todo.
Indeed I wanted to add a ‘to_matrix()’ to get the dense matrix but I forgot. 
this should be easy to add.

for the call... the right operation would be \ (inverse division), or division of an array (the F) by S.
Since I think that does not exist in Python then i guess solve is a better name